x/capability/keeper/keeper.go

```diff
@@ -0,0 +1,244 @@
+		}
+	}
+
+	k.sealed = true
+}
+
+// NewCapability attempts to create a new capability with a given name. If the
+// capability already exists in the in-memory store, an error will be returned.
+// Otherwise, a new capability is created with the current global index. The
+// newly created capability has the module and name tuple set as the initial owner.
+// Finally, the global index is incremented along with forward and reverse indexes
+// set in the in-memory store.
+func (sk ScopedKeeper) NewCapability(ctx sdk.Context, name string) (types.Capability, error) {
+	store := ctx.KVStore(sk.storeKey)
+	memStore := ctx.KVStore(sk.memKey)
+
+	if memStore.Has(types.RevCapabilityKey(sk.module, name)) {
+		return nil, sdkerrors.Wrapf(types.ErrCapabilityTaken, fmt.Sprintf("module: %s, name: %s", sk.module, name))
+	}
+
+	// create new capability with the current global index
+	index := types.IndexFromKey(store.Get(types.KeyIndex))
+	cap := types.NewCapabilityKey(index)
+
+	// update capability owner set
+	if err := sk.addOwner(ctx, cap, name); err != nil {
+		return nil, err
+	}
+
+	// increment global index
+	store.Set(types.KeyIndex, types.IndexToKey(index+1))
+
+	// Set the forward mapping between the module and capability tuple and the
+	// capability name in the in-memory store.
+	memStore.Set(types.FwdCapabilityKey(sk.module, cap), []byte(name))
+
+	// Set the reverse mapping between the module and capability name and the
+	// capability in the in-memory store.
+	memStore.Set(types.RevCapabilityKey(sk.module, name), sk.cdc.MustMarshalBinaryBare(cap))
+
+	logger(ctx).Info("created new capability", "module", sk.module, "name", name)
+	return cap, nil
+}
+
+// AuthenticateCapability attempts to authenticate a given capability and name
+// from a caller. It allows for a caller to check that a capability does in fact
+// correspond to a particular name. The scoped keeper will lookup the capability
+// from the internal in-memory store and check against the provided name. It returns
+// true upon success and false upon failure.
+//
+// Note, the capability's forward mapping is indexed by a string which should
+// contain it's unique memory reference.
+func (sk ScopedKeeper) AuthenticateCapability(ctx sdk.Context, cap types.Capability, name string) bool {
+	memStore := ctx.KVStore(sk.memKey)
+
+	bz := memStore.Get(types.FwdCapabilityKey(sk.module, cap))
+	return string(bz) == name
+}
+
+// ClaimCapability attempts to claim a given Capability and name tuple. This tuple
+// is considered an Owner. It will attempt to add the owner to the persistent
+// set of capability owners for the capability index. If the owner already exists,
+// it will return an error. Otherwise, it will also set a forward and reverse index
+// for the capability and capability name.
+func (sk ScopedKeeper) ClaimCapability(ctx sdk.Context, cap types.Capability, name string) error {
+	// update capability owner set
+	if err := sk.addOwner(ctx, cap, name); err != nil {
```

Comment:
Should we check that capability already exists? If not, can this replace functionality of `NewCapability`?

Reply:
Hmm, I think it is better to keep them separate; when calling `ClaimCapability` the calling module does not intend to create a new capability.